Basic support for Confluent's  KafkaAvroSerializer / KafkaAvroDeSerializer

fmck3516 opened the pull request:
Basic support to serialize and deserialize with Confluent's [proprietary 5-byte prefix](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format).

The prefix will be added if the key or value serializer is set to `io.confluent.kafka.serializers.KafkaAvroSerializer`. 
The prefix will be removed if the key or value deserializer is set to `io.confluent.kafka.serializers.KafkaAvroDeserializer`.

**Configuration example:**
```
var configuration = JSON.stringify({
    consumer: {
        keyDeserializer: "io.confluent.kafka.serializers.KafkaAvroDeserializer",
        valueDeserializer: "io.confluent.kafka.serializers.KafkaAvroDeserializer",
    },
    producer: {
        keySerializer: "io.confluent.kafka.serializers.KafkaAvroSerializer",
        valueSerializer: "io.confluent.kafka.serializers.KafkaAvroSerializer",
    },
    schemaRegistry: {
        url: "https://subdomain.us-east-2.aws.confluent.cloud",
        basicAuth: {
            credentialsSource: "USER_INFO",
            userInfo: "KEY:SECRET"
        },
    },
})
```

The configuration for `schemaRegistry` is required to use a serializer of type `io.confluent.kafka.serializers.KafkaAvroSerializer`.

A new `produce` and `consume` function has been added (each accepts the new configuration argument):

`produceWithConfiguration(producer, messages, configuration, keySchema, valueSchema);`
`consumeWithConfiguration(consumer, 20, configuration, keySchema, valueSchema);`

**Note**: I ported a sub-set of the existing functionality of the Confluent Java client. This code will be obsolete once Confluent adds Schema Registry support for their Go client.  At the time of writing, the Go client does not support Schema Registry: https://docs.confluent.io/platform/current/clients/index.html

fmck3516 commented:
I'm not familiar with Go - I still have to figure out how to unit test the stuff.

mostafa commented on test/test.js:
```diff
@@ -0,0 +1,63 @@
+/*
```
I am wondering if you could move all the scripts (`test_*.js`) to the `scripts` directory and rename this somehow to something like `test_avro_with_magic_bytes.js` or something shorter.

fmck3516 replied:
I renamed the file to `test_avro_with_schema_registry.js` and yes, I think it would be good to move the tests out of the top level directory.

mostafa replied:
But the file is still called `test/test.js`. Am I missing something? :thinking:

mostafa reviewed (changes requested):
Hey @fmck3516,

Thanks for the awesome PR. :pray: I left a few comments and I think it can be merged after the changes.

fmck3516 commented:
@mostafa: Thanks for the feedback - I will update the PR in the upcoming week.

fmck3516 commented on consumer.go:
```diff
@@ -88,16 +101,18 @@ func (*Kafka) Consume(
 		message := make(map[string]interface{})
 		if len(msg.Key) > 0 {
-			message["key"] = string(msg.Key)
+			keyWithoutPrefix := removeMagicByteAdnSchemaIdPrefix(properties, msg.Key, "key")
```
Typo - should be `And`. 

> Could it be just `removeMagicByte`?

No, that would be technically incorrect. The magic byte is the leading zero. I need to remove the following 4-bytes for the schema id as well.

fmck3516 commented on schemaRegistry.go:
```diff
@@ -0,0 +1,91 @@
+	if schemaId != 0 {
+		return append(append([]byte{ 0 }, i32tob(schemaId)...), avroData...), nil
+	}
+	return avroData, nil
```
I'd rather fail fast in this case. The schema lookup is only performed if the serializer is set to `io.confluent.kafka.serializers.KafkaAvroSerializer`. A message without prefix would violate Confluent's wire format contract.

fmck3516 commented:
I incorporated your feedback. I also switched to a JSON-based configuration model to be in sync with the approach for the `auth` configuration.

What do you think about having one big configuration JSON for auth and everything else?

fmck3516 commented on producer.go:
```diff
@@ -60,9 +79,19 @@ func (*Kafka) Produce(
 			value = ToAvro(message["value"], valueSchema)
 		}
 
+		keyData, err := addMagicByteAdnSchemaIdPrefix(properties, key, kafkaMessages[i].Topic, "key", keySchema)
```
Typo - should be `And`. 

> Could it be just `addMagicByte`?

No, that would be technically incorrect. The magic byte is only the leading zero. I need to add the 4-bytes for the schema id as well.

mostafa commented:
@fmck3516 I was actually thinking about the same thing, but then thought that it's out of scope of this PR or mine. So maybe the next PR would focus on refactoring, cleanups and better configuration management. I also have the idea to refactor metric collection and error handling. Sorry for too much information! :rofl:

mostafa commented on test_avro_with_schema_registry.js:
```diff
@@ -0,0 +1,102 @@
+}`
+
+
+var configuration = JSON.stringify({
```
I love this. :clap:

mostafa reviewed (approved):
LGTM! I can rearrange test scripts (files) after merging.

fmck3516 commented:
> @fmck3516 I was actually thinking about the same thing, but then thought that it's out of scope of this PR or mine.
That makes sense - one step at a time. 

>  So maybe the next PR would focus on refactoring, cleanups and better configuration management. I also have the idea to refactor metric collection and error handling. Sorry for too much information! 🤣
One thing I would like to add to the wishlist: A suite of regression tests that is executed automatically.
